kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,305 @@
+                let f = |t| {
+                    let Term {
+                        exp: exp2,
+                        sign: sign2,
+                        degree: degree2,
+                    } = t;
+                    let degree = &degree * &degree2;
+                    let sign = sign != sign2;
+                    //TODO: specialize square case
+                    let exp = FoldingExp::Mul(exp.clone(), Box::new(exp2));
+                    Term { exp, sign, degree }
+                };
+                let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+                extract_terms(*e.clone(), terms, &f);
+            }
+        }
+        FoldingExp::Add(e1, e2) => {
+            extract_terms(*e1, terms, f);
+            extract_terms(*e2, terms, f);
+        }
+        FoldingExp::Sub(e1, e2) => {
+            extract_terms(*e1, terms, f);
+            let f = |t: Term<C>| f(-t);
+            let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+            extract_terms(*e2, terms, &f);
+        }
+        FoldingExp::Mul(e1, e2) => {
+            let mut temp = Vec::with_capacity(4);
+            extract_terms(*e1, &mut temp, f);
+            for term in temp {
+                let Term { exp, sign, degree } = term;
+                let exp = Box::new(exp);
+                let f = |t| {
+                    let Term {
+                        exp: exp2,
+                        sign: sign2,
+                        degree: degree2,
+                    } = t;
+                    let degree = &degree * &degree2;
+                    let sign = sign == sign2;
+                    let exp = FoldingExp::Mul(exp.clone(), Box::new(exp2));
+                    Term { exp, sign, degree }
+                };
+                let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+                extract_terms(*e2.clone(), terms, &f);
+            }
+        }
+        e @ FoldingExp::UnnormalizedLagrangeBasis(_) => {
+            let term = f(Term::new(e, Zero));
+            terms.push(term)
+        }
+        FoldingExp::Cache(id, e) => {
+            let f = |t| {
+                let Term { exp, sign, degree } = t;
+                let exp = FoldingExp::Cache(id, Box::new(exp));
+                Term { exp, sign, degree }
+            };
+            let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+            extract_terms(*e, terms, &f);
+        }
+    }
+}
+
+///simplyfies the expression, and separates it into terms of different degree y
+pub fn folding_expression<C: Copy, F: Fn(&FeatureFlag) -> bool>(
+    exp: &Expr<C>,
+    flag_resolver: &F,
+) -> IntegratedFoldingExpr<C> {
+    let simplied = simplify_expression(exp, flag_resolver);
+    let mut terms = vec![];
+    extract_terms(simplied, &mut terms, &|t| t);
+    let mut integrated = IntegratedFoldingExpr::default();
+    for term in terms.into_iter() {
+        let Term { exp, sign, degree } = term;
+        let t = (exp, sign);
+        match degree {
```

Comment:
Stupid question, but: isn't it a problem that `integrated` is now sorted differently than `terms`. Does the order in which the terms are extracted matter at all?

Reply:
I think it isn't an issue as long as `folding_expression` is pure, and even if the order where to be different each time, the expression should evaluate to the same.

Reply:
in some way it is the actual polynomial, the order shouldn't matter, at some point it will have to be turned into `Expr` again to provide it to the proof system, and I'll change it later so that it is handled by folding itself and you don't have to know `IntegratedFoldingExpr` even exists.

Reply:
what do you mean by "pure"?